File: chunks/term_diff/lib/term_diff/git/runner.ex
defmodule TermDiff.Git.Runner do
  @moduledoc "Execute git commands via Erlang Port. Imperative shell."

  @git_timeout 15_000

  @spec status(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def status(repo_path) do
    run(repo_path, ["status", "--porcelain=v1", "-u"])
  end

  @spec plumbing_status(String.t()) :: {:ok, [TermDiff.Diff.FileState.t()]} | {:error, String.t()}
  def plumbing_status(repo_path) do
    with {:ok, staged_raw} <- run(repo_path, ["diff-index", "--cached", "--name-status", "HEAD"]),
         {:ok, unstaged_raw} <- run(repo_path, ["diff-files", "--name-status"]),
         {:ok, untracked_raw} <- run(repo_path, ["ls-files", "--others", "--exclude-standard"]) do
      staged = parse_name_status(staged_raw)
      unstaged = parse_name_status(unstaged_raw)
      untracked = parse_untracked(untracked_raw)

      files = merge_file_states(staged, unstaged, untracked)
      {:ok, files}
    end
  end

  defp parse_name_status(""), do: %{}

  defp parse_name_status(raw) do
    raw
    |> String.split("\n", trim: true)
    |> Map.new(fn line ->
      case String.split(line, "\t", parts: 2) do
        [status, path] -> {path, status_letter(status)}
        _ -> {line, :unknown}
      end
    end)
  end

  defp parse_untracked(""), do: []
  defp parse_untracked(raw), do: String.split(raw, "\n", trim: true)

  defp status_letter("M"), do: :modified
  defp status_letter("A"), do: :added
  defp status_letter("D"), do: :deleted
  defp status_letter("R" <> _), do: :renamed
  defp status_letter(_), do: :modified

  defp merge_file_states(staged, unstaged, untracked) do
    all_paths =
      (Map.keys(staged) ++ Map.keys(unstaged) ++ untracked)
      |> Enum.uniq()
      |> Enum.sort()

    Enum.map(all_paths, fn path ->
      s = Map.get(staged, path)
      u = Map.get(unstaged, path)
      is_untracked = path in untracked

      status =
        cond do
          is_untracked -> :untracked
          s == :added && u == :modified -> :partial_modified
          s == :added -> :staged_new
          s == :modified && u == :modified -> :partial_modified
          s == :modified -> :staged_modified
          s == :deleted -> :staged_deleted
          s == :renamed && u == :modified -> :partial_modified
          s == :renamed -> :staged_renamed
          u == :modified -> :unstaged_modified
          u == :deleted -> :unstaged_deleted
          true -> :untracked
        end

      %TermDiff.Diff.FileState{path: path, status: status}
    end)
  end

  @spec diff(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def diff(repo_path) do
    run(repo_path, ["diff"])
  end

  @spec diff_staged(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def diff_staged(repo_path) do
    run(repo_path, ["diff", "--staged"])
  end

  @spec branch(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def branch(repo_path) do
    # rev-parse is already plumbing
    run(repo_path, ["rev-parse", "--abbrev-ref", "HEAD"])
  end

  @spec log_oneline(String.t(), non_neg_integer()) :: {:ok, String.t()} | {:error, String.t()}
  def log_oneline(repo_path, count \\ 50) do
    run(repo_path, ["log", "--oneline", "--no-decorate", "-#{count}"])
  end

  @spec log_show(String.t(), String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def log_show(repo_path, commit_hash) do
    run(repo_path, ["show", commit_hash, "--stat", "--format=full"])
  end

  @spec log_diff(String.t(), String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def log_diff(repo_path, commit_hash) do
    run(repo_path, ["show", commit_hash, "--format="])
  end

  @spec stage(String.t(), String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def stage(repo_path, file_path) do
    run(repo_path, ["add", file_path])
  end

  @spec unstage(String.t(), String.t(), keyword()) :: {:ok, String.t()} | {:error, String.t()}
  def unstage(repo_path, file_path, opts \\ []) do
    staged_status = Keyword.get(opts, :staged_status)

    if staged_status == :added do
      run(repo_path, ["rm", "--cached", file_path])
    else
      run(repo_path, ["restore", "--staged", file_path])
    end
  end

  @spec exec_file_command(String.t(), TermDiff.Diff.FileState.command()) ::
          {:ok, String.t()} | {:error, String.t()} | :noop
  def exec_file_command(_repo_path, :noop), do: :noop
  def exec_file_command(repo_path, {:add, path}), do: run(repo_path, ["add", path])

  def exec_file_command(repo_path, {:rm_cached, path}),
    do: run(repo_path, ["rm", "--cached", path])

  def exec_file_command(repo_path, {:restore_staged, path}),
    do: run(repo_path, ["restore", "--staged", path])

  @spec diff_untracked(String.t(), String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def diff_untracked(repo_path, file_path) do
    run(repo_path, ["diff", "--no-index", "/dev/null", file_path], expected_exits: [0, 1])
  end

  @spec commit(String.t(), String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def commit(repo_path, message) do
    run(repo_path, ["commit", "-m", message])
  end

  @spec commit_amend(String.t(), String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def commit_amend(repo_path, message) do
    run(repo_path, ["commit", "--amend", "-m", message])
  end

  @spec last_commit_message(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def last_commit_message(repo_path) do
    run(repo_path, ["log", "-1", "--format=%B"])
  end

  @spec open_editor(String.t(), String.t()) :: :ok | {:error, String.t()}
  def open_editor(repo_path, file_path) do
    editor = System.get_env("EDITOR", "vim")
    full_path = Path.join(repo_path, file_path)

    spawn(fn ->
      System.cmd(editor, [full_path], into: IO.stream())
    end)

    :ok
  end

  # ── Port-based execution ──

  @spec run(String.t(), [String.t()], keyword()) :: {:ok, String.t()} | {:error, String.t()}
  defp run(repo_path, args, opts \\ []) do
    expected_exits = Keyword.get(opts, :expected_exits, [0])
    git = System.find_executable("git")

    port =
      Port.open({:spawn_executable, git}, [
        :binary,
        :exit_status,
        :stderr_to_stdout,
        args: args,
        cd: repo_path
      ])

    collect_output(port, [], expected_exits)
  rescue
    ErlangError ->
      {:error, "failed to execute git"}
  end

  defp collect_output(port, acc, expected_exits) do
    receive do
      {^port, {:data, data}} ->
        collect_output(port, [data | acc], expected_exits)

      {^port, {:exit_status, code}} ->
        output = acc |> Enum.reverse() |> IO.iodata_to_binary() |> String.trim_trailing()

        if code in expected_exits do
          {:ok, output}
        else
          {:error, output}
        end
    after
      @git_timeout ->
        Port.close(port)
        {:error, "git command timed out"}
    end
  end
end

File: chunks/blimp_diff_cpp/src/git/runner.cpp
#include "git/runner.h"
#include <array>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <stdexcept>
#include <sstream>

namespace blimp::git {

Runner::Runner(std::filesystem::path repo_root)
    : root_(std::move(repo_root)) {}

CmdResult Runner::run(const std::vector<std::string>& args,
                       bool allow_nonzero) const {
    std::ostringstream cmd;
    cmd << "cd " << root_.string() << " && git";
    for (const auto& arg : args) {
        cmd << " ";
        // Always single-quote args to handle spaces, newlines, special chars.
        // Single quotes protect everything except single quotes themselves.
        cmd << "'";
        for (char c : arg) {
            if (c == '\'') cmd << "'\\''";
            else cmd << c;
        }
        cmd << "'";
    }
    cmd << " 2>&1";

    CmdResult result;
    std::array<char, 4096> buffer{};
    FILE* pipe = popen(cmd.str().c_str(), "r");
    if (!pipe) {
        result.exit_code = -1;
        result.stderr_str = "Failed to open pipe";
        return result;
    }
    while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
        result.stdout_str += buffer.data();
    }
    int raw = pclose(pipe);
    result.exit_code = WEXITSTATUS(raw);

    if (result.exit_code != 0 && !allow_nonzero) {
        result.stderr_str = result.stdout_str;
    }
    return result;
}

CmdResult Runner::status() const {
    return run({"status", "--porcelain=v1"});
}

CmdResult Runner::diff() const {
    return run({"diff"});
}

CmdResult Runner::diff_staged() const {
    return run({"diff", "--staged"});
}

CmdResult Runner::diff_untracked(const std::string& path) const {
    return run({"diff", "--no-index", "/dev/null", path}, /*allow_nonzero=*/true);
}

CmdResult Runner::branch() const {
    return run({"branch", "--show-current"});
}

CmdResult Runner::log_oneline(int count) const {
    return run({"log", "--oneline", "-n", std::to_string(count)});
}

CmdResult Runner::log_show(const std::string& hash) const {
    return run({"show", "--stat", hash});
}

CmdResult Runner::log_diff(const std::string& hash) const {
    return run({"show", "--format=", hash});
}

CmdResult Runner::stage(const std::string& path) const {
    return run({"add", path});
}

CmdResult Runner::unstage_restore(const std::string& path) const {
    return run({"restore", "--staged", path});
}

CmdResult Runner::unstage_rm_cached(const std::string& path) const {
    return run({"rm", "--cached", path});
}

CmdResult Runner::commit(const std::string& message) const {
    return run({"commit", "-m", message});
}

CmdResult Runner::commit_amend(const std::string& message) const {
    return run({"commit", "--amend", "-m", message});
}

void Runner::open_editor(const std::string& path) const {
    const char* editor = std::getenv("EDITOR");
    if (!editor) editor = "vim";
    std::string cmd = std::string(editor) + " " + (root_ / path).string() + " &";
    std::system(cmd.c_str());
}

} // namespace blimp::git
